In [1]:
# импортируем библиотеки

import pandas as pd
from sqlalchemy import create_engine

# устанавливаем параметры

db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Вводные данные

**Канал: сервис для чтения книг по подписке**

**Задача: проанализировать базу данных**

**Дано:  информация о книгах, издательствах, авторах, а также пользовательские обзоры книг**

### Цели и задачи

**Цель:** Провести анализ базы данных. 

**Задачи :**

1. Посчитайте, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитайте количество обзоров и среднюю оценку;
3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа - брошюры;
4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Ознакомление с данными

Датасет **`books`** - иноформация о **книгах**:

Датасет **`authors`** - иноформация об **авторах**:

Датасет **`publishers`** - иноформация об **издательствах**:

Датасет **`ratings`** - иноформация об **пользовательских оценках книг**:

Датасет **`reviews`** - иноформация о **пользовательских обзорах**:



#### Структура **books**

**`book_id`** — идентификатор книги;

**`author_id`** — идентификатор автора;

**`title`** — название книги;

**`num_pages`** — количество страниц;

**`publication_date`** — дата публикации книги;

**`publisher_id`** — идентификатор издателя.



#### Структура **authors**

**`author_id`** — идентификатор автора;

**`author`** — имя автора.


#### Структура **publishers**

**`publisher_id`** — идентификатор издательства;

**`publisher`** — название издательства;



#### Структура **ratings**

**`rating_id`** — идентификатор оценки;

**`book_id`** — идентификатор книги;

**`username`** — имя пользователя, оставившего оценку;

**`rating`** — оценка книги.


#### Структура **reviews**

**`review_id`** — идентификатор обзора;

**`book_id`** — идентификатор книги;

**`username`** — имя автора обзора;

**`text`** — текст обзора.


### Схема данных

<img src="https://pictures.s3.yandex.net/resources/scheme_1589269096.png"/>

### Ход работы

1. [x] Описать цели исследования
2. [x] Обзор данных: Вывести первые строки
3. [x] Исследование данных:Сделать по одному SQL-запросу для решения каждого задания
4. [x] Вывести результаты каждого запроса в тетрадке
5. [x] Описать выводы по каждой из решённых задач.
6. [x] Общее заключение


## Обзор данных: вывести первые строки

In [2]:
def first_look (data):
    print('---------Пропуски-----------\n')
    display(data.isna().sum())
    print('\n---------Процент пропусков от всего датасета-----------')
    display(data.isna().mean().sort_values(ascending=False)*100)
    print('\n---------Дубликаты-----------\n')
    if data.duplicated().sum() > 0:
        print('Дубликатов: ', data.duplicated().sum())
    else:
        print('Дубликатов НЕТ')

перевод в переменные для пандас

In [3]:
query = ''' 
SELECT *
FROM books
''' 
books = pd.io.sql.read_sql(query, con = engine)
query = ''' 
SELECT *
FROM authors
''' 
authors = pd.io.sql.read_sql(query, con = engine)
query = ''' 
SELECT *
FROM publishers
''' 
publishers = pd.io.sql.read_sql(query, con = engine)
query = ''' 
SELECT *
FROM ratings
''' 
ratings = pd.io.sql.read_sql(query, con = engine)
query = ''' 
SELECT *
FROM reviews
''' 
reviews = pd.io.sql.read_sql(query, con = engine)

In [4]:
first_look(books)
first_look(authors)
first_look(publishers)
first_look(ratings)
first_look(reviews)

---------Пропуски-----------



book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64


---------Процент пропусков от всего датасета-----------


book_id             0.0
author_id           0.0
title               0.0
num_pages           0.0
publication_date    0.0
publisher_id        0.0
dtype: float64


---------Дубликаты-----------

Дубликатов НЕТ
---------Пропуски-----------



author_id    0
author       0
dtype: int64


---------Процент пропусков от всего датасета-----------


author_id    0.0
author       0.0
dtype: float64


---------Дубликаты-----------

Дубликатов НЕТ
---------Пропуски-----------



publisher_id    0
publisher       0
dtype: int64


---------Процент пропусков от всего датасета-----------


publisher_id    0.0
publisher       0.0
dtype: float64


---------Дубликаты-----------

Дубликатов НЕТ
---------Пропуски-----------



rating_id    0
book_id      0
username     0
rating       0
dtype: int64


---------Процент пропусков от всего датасета-----------


rating_id    0.0
book_id      0.0
username     0.0
rating       0.0
dtype: float64


---------Дубликаты-----------

Дубликатов НЕТ
---------Пропуски-----------



review_id    0
book_id      0
username     0
text         0
dtype: int64


---------Процент пропусков от всего датасета-----------


review_id    0.0
book_id      0.0
username     0.0
text         0.0
dtype: float64


---------Дубликаты-----------

Дубликатов НЕТ


Дубликатов в датасетах необнаружено

### books

Первые пять строчек books

In [5]:
query = ''' 
SELECT *
FROM books
LIMIT 5;
''' 
books = pd.io.sql.read_sql(query, con = engine)

Количество книг

In [6]:
query = '''
SELECT 
    COUNT(book_id)
FROM Books;
''' 
pd.io.sql.read_sql(query, con = engine)

count
0   1000

В датасете хранится информация о 1000 книг

### authors

Первые пять строчек authors

In [7]:
query = '''
SELECT *
FROM authors
LIMIT 5;
''' 
pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Количество авторов

In [8]:
query = '''
SELECT 
    COUNT(author_id)
FROM authors;
''' 
pd.io.sql.read_sql(query, con = engine)

count
0    636

В датасете хранится информация о 636 авторах. 

### publishers

Первые пять строчек publishers

In [9]:
query = '''
SELECT *
FROM publishers
LIMIT 5;
''' 
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Количество издательств

In [10]:
query = '''
SELECT 
    COUNT(publisher_id)
FROM publishers;
''' 
pd.io.sql.read_sql(query, con = engine)

count
0    340

В датасете хранится информация о 340 издательствах.

### ratings

Первые пять строчек ratings

In [11]:
query = '''
SELECT *
FROM ratings
LIMIT 5;
''' 
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Количество пользовательских оценок книг

In [12]:
query = '''
SELECT 
    COUNT(rating_id)
FROM ratings;
''' 
pd.io.sql.read_sql(query, con = engine)

count
0   6456

В датасете хранится 6456 пользовательских оценок про книги

### reviews

Первые пять строчек reviews

In [13]:
query = '''
SELECT *
FROM reviews
LIMIT 5;
''' 
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Количество пользовательских обзоров

In [14]:
query = '''
SELECT 
    COUNT(review_id)
FROM reviews;
''' 
pd.io.sql.read_sql(query, con = engine)

count
0   2793

В датасете хранится 2793 пользовательских обзоров про книги

### Вывод

- Дубликатов в датасетах необнаружено
- В датасете **`books`** хранится информация о 1000 книг
- В датасете **`authors`** хранится информация о 636 авторах.
- В датасете **`publishers`** хранится информация о 340 издательствах.
- В датасете **`ratings`** хранится 6456 пользовательских оценок про книги
- В датасете **`reviews`** хранится 2793 пользовательских обзоров про книги

## Исследование данных: Сделать по одному SQL-запросу для решения каждого задания

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [16]:
query = '''
SELECT 
    COUNT(*)
FROM books
WHERE publication_date > '2000-01-01';

'''
pd.io.sql.read_sql(query, con = engine)

count
0    819

После 1 января 2000 года вышло 819 книг. 

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [25]:
query = '''
SELECT 
    books.title AS book_title,
    COUNT(DISTINCT(reviews.text)) AS amount_reviews,
    AVG(ratings.rating) AS avg_rating
FROM books
LEFT JOIN ratings ON ratings.book_id = books.book_id
LEFT JOIN reviews ON reviews.book_id = books.book_id
GROUP BY books.book_id
ORDER BY avg_rating DESC, amount_reviews DESC, book_title ASC 
'''

pd.io.sql.read_sql(query, con = engine)



book_title  amount_reviews  \
0                        A Dirty Job (Grim Reaper  #1)               4   
1         Moneyball: The Art of Winning an Unfair Game               3   
2              School's Out—Forever (Maximum Ride  #2)               3   
3                A Fistful of Charms (The Hollows  #4)               2   
4            A Woman of Substance (Emma Harte Saga #1)               2   
..                                                 ...             ...   
995  The World Is Flat: A Brief History of the Twen...               3   
996                                      Drowning Ruth               3   
997                  His Excellency: George Washington               2   
998                                              Junky               2   
999                               Harvesting the Heart               2   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 3 columns]

Получили название книги, количество обзоров (по убыванию) и среднюю оценку (по убыванию). 

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа - брошюры

In [26]:
query = '''
SELECT 
    p.publisher_id AS publisher_id,
    p.publisher AS publisher_name,
    COUNT(b.book_id) AS amount_books
FROM books AS b
LEFT JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50 
GROUP BY p.publisher_id
ORDER BY amount_books DESC
LIMIT 1
'''
pd.io.sql.read_sql(query, con = engine)

publisher_id publisher_name  amount_books
0           212  Penguin Books            42

Издательство Penguin Books выпустило наибольшее число книг, которые толще 50 страниц

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [27]:
query = '''
SELECT b.author_id,
       a.author,
       ROUND(AVG(rate.rating),2) AS avg_rating,
       COUNT(b.book_id) AS amount_books
FROM books AS b
LEFT JOIN ratings AS rate ON b.book_id = rate.book_id
LEFT JOIN authors AS a ON b.author_id = a.author_id
WHERE b.book_id in 
    (SELECT b.book_id
    FROM books AS b 
    LEFT JOIN ratings AS rate ON b.book_id = rate.book_id 
    GROUP BY b.book_id 
    HAVING COUNT(DISTINCT rate.rating_id) >=50)
GROUP BY b.author_id, author
ORDER BY avg_rating DESC
LIMIT 3
''' 
pd.io.sql.read_sql(query, con = engine)

author_id                             author  avg_rating  amount_books
0        236         J.K. Rowling/Mary GrandPré        4.29           310
1        402  Markus Zusak/Cao Xuân Việt Khương        4.26            53
2        240                     J.R.R. Tolkien        4.25           162

Немного непонятно как толковать вопрос, опираясь на полученные данные. В вопросе спрашивается про автора, то есть в единственном числе. Тогда ТОП-2 не подходит и ответом будет Джон Рональд Руэл Толкин.

Но если брать author_id как единого автора, то ответом будет Джоан Роулинг/Мэри Гранпрэ.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [ ]:
query = '''
SELECT 
    ROUND(AVG(rate.amount_rev),2) AS avg_rating
FROM (SELECT 
        username,
        COUNT (DISTINCT review_id) AS amount_rev
      FROM reviews
      WHERE username in (SELECT 
                              username
                         FROM ratings
                         GROUP BY username
                         HAVING COUNT (DISTINCT rating_id) > 50)
      GROUP BY username) AS rate
''' 
pd.io.sql.read_sql(query, con = engine)

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок равняется **24**. 

## Общее заключение

1. Про датасеты в целом:
    - В датасете **`books`** хранится информация о 1000 книг
    - В датасете **`authors`** хранится информация о 636 авторах.
    - В датасете **`publishers`** хранится информация о 340 издательствах.
    - В датасете **`ratings`** хранится 6456 пользовательских оценок про книги
    - В датасете **`reviews`** хранится 2793 пользовательских обзоров про книги
2. По задачам:
    - После **1 января 2000** года вышло **819** книг
    - **ТОП-1** **A Dirty Job (Grim Reaper #1)** - Средняя оценка **5.0**, кол-во обзоров **4**. 
    - Издательство **Penguin Books** выпустило **наибольшее число** книг (**42**), которые **толще 50 страниц**
    - Автор с самой высокой средней оценкой книг — **Джоан Роулинг/Мэри Гранпрэ** (более 50 оценок). Средний рейтинг **4.29** по 310 книгам. 
    - **Среднее количество обзоров** от пользователей, которые поставили **больше 50 оценок** равняется **24**.